lib/AST/GenericSignature.cpp

```diff
@@ -626,15 +626,29 @@ unsigned GenericSignatureImpl::getGenericParamOrdinal(
 }
 
 Type GenericSignatureImpl::getNonDependentUpperBounds(Type type) const {
+  return getUpperBound(type);
+}
+
+Type GenericSignatureImpl::getDependentUpperBounds(Type type) const {
+  return getUpperBound(type, /*wantNonDependentBound=*/false);
+}
+
+Type GenericSignatureImpl::getUpperBound(Type type,
+                                         bool wantDependentBound) const {
   assert(type->isTypeParameter());
 
   bool hasExplicitAnyObject = requiresClass(type);
 
   llvm::SmallVector<Type, 2> types;
+
+  // If the class contains a type parameter that cannot be reduced,
+  // try looking for a non-dependent superclass.
   if (Type superclass = getSuperclassBound(type)) {
-    // If the class contains a type parameter, try looking for a non-dependent
-    // superclass.
-    while (superclass && superclass->hasTypeParameter()) {
+    while (superclass->hasTypeParameter()) {
+      superclass = getReducedType(superclass);
+      if (wantDependentBound && !superclass->hasTypeParameter()) {
```

Comment:
I think this should be `if (wantDependentBound || !superclass->hasTypeParameter()) break;`

Reply:
`if (wantDependentBound || !superclass->hasTypeParameter())` causes unhandled coercion error because it will try to erase the type to some value with an unresolved dependent member

Reply:
@angela-laar when you flipped the check you didn't adjust it's use, that might be the reason why this is happening.